crates/standalone/src/lib.rs

```diff
@@ -487,87 +481,150 @@ impl StandaloneEnv {
+
+                if let Some(hash) = maybe_hash {
+                    ensure!(
+                        hash == database.program_bytes_address,
+                        "database already initialized with module {} (requested: {})",
+                        hash,
+                        database.program_bytes_address,
+                    );
+                    if !self.host_controller.has_module_host(&ctx) {
+                        log::info!("Re-spawing database (module: {})", hash);
+                        self.host_controller.spawn_module_host(ctx).await?;
+                    } else {
+                        log::info!("Database already initialized with module {}", hash);
+                    }
+                } else {
+                    self.host_controller.init_module_host(lock.token() as u128, ctx).await?;
+                }
+
+                Ok(())
             }
-            Ok(())
-        } else {
-            // Start and init the service
-            let mut state = DatabaseInstanceState {
-                database_instance_id: instance.id,
-                initialized: false,
-            };
-            state.initialized = true;
-            self.worker_db.upsert_database_instance_state(state.clone()).unwrap();
-            self.init_module_on_database_instance(instance.database_id, instance.id)
-                .await?;
-            self.worker_db.upsert_database_instance_state(state).unwrap();
-            Ok(())
         }
     }
 
     async fn on_update_database_instance(
         &self,
         instance: &DatabaseInstance,
     ) -> Result<Option<UpdateDatabaseResult>, anyhow::Error> {
-        let state = self.worker_db.get_database_instance_state(instance.id)?;
-        match state {
-            Some(state) if state.initialized => self
-                .update_module_on_database_instance(instance.database_id, instance.id)
-                .await
-                .map(Some),
-
-            _ => self.on_insert_database_instance(instance).await.map(|()| None),
+        let database = self
+            .control_db
+            .get_database_by_id(instance.database_id)?
+            .ok_or_else(|| {
+                anyhow!(
+                    "unknown database: id: {}, instance: {}",
+                    instance.database_id,
+                    instance.id
+                )
+            })?;
+
+        let ctx = self.load_module_host_context(database.clone(), instance.id).await?;
+        let stdb = &ctx.dbic.relational_db;
+        let tx = stdb.begin_tx();
+
+        match stdb.program_hash(&tx) {
+            Err(e) => {
+                stdb.rollback_tx(tx);
+                Err(e.into())
+            }
+
+            Ok(maybe_hash) => {
+                // Release tx due to locking semantics and acquire a control db
+                // lock instead.
+                stdb.commit_tx(tx)?;
+                let lock = self.lock_database_instance_for_update(instance.id)?;
```

Comment:
This feels like a race condition...

Reply:
It is, but we can’t thread through the tx. If a later thread is faster acquiring the lock, this update will not have any effect — that is, we guarantee at most once, not exactly once.

Reply:
Ah, ok. Not ideal, but probably fine.